sdk/core/core-lro/samples/v3/javascript/customPoller.js

```diff
@@ -0,0 +1,175 @@
+    setDelay(response.retryAfter); // updates the delay if the service suggests a different interval
+  }
+}
+
+function deserializeState(serializedState) {
+  try {
+    return JSON.parse(serializedState).state;
+  } catch (e) {
+    throw new Error(`Unable to deserialize input state: ${serializedState}`);
+  }
+}
+
+function createPoller({ intervalInMs, restoreFrom }) {
+  let statePromise;
+  let state;
+  if (restoreFrom) {
+    state = deserializeState(restoreFrom);
+    statePromise = Promise.resolve(state);
+  } else {
+    statePromise = initOperation().then((s) => (state = s));
+  }
+  let resultPromise;
+  const abortController = new AbortController();
+  const handlers = new Map();
+  const handleProgressEvents = async () => handlers.forEach((h) => h(state));
+  const cancelErrMsg = "Operation was canceled";
+  let currentPollIntervalInMs = intervalInMs ?? DEFAULT_POLL_INTERVAL_IN_MS;
+  const poller = {
+    get operationState() {
+      return state;
+    },
+    get result() {
+      return state?.result;
+    },
+    get isDone() {
+      return ["succeeded", "failed", "canceled"].includes(state?.status ?? "");
+    },
+    onProgress: (callback) => {
+      const s = Symbol();
+      handlers.set(s, callback);
+      return () => handlers.delete(s);
+    },
+    serialize: async () => {
+      await statePromise;
+      return JSON.stringify({
+        state,
+      });
+    },
+    submitted: async () => {
+      await statePromise;
+    },
+    pollUntilDone: async (pollOptions) => {
+      resultPromise ??= (async () => {
+        await statePromise;
+        if (!state) {
+          throw new Error("Poller should be initialized but it is not!");
+        }
+        const { abortSignal: inputAbortSignal } = pollOptions || {};
+        // In the future we can use AbortSignal.any() instead
+        function abortListener() {
+          abortController.abort();
+        }
+        const abortSignal = abortController.signal;
+        if (inputAbortSignal?.aborted) {
+          abortController.abort();
+        } else if (!abortSignal.aborted) {
+          inputAbortSignal?.addEventListener("abort", abortListener, { once: true });
+        }
+
+        try {
+          if (!poller.isDone) {
+            await poller.poll({ abortSignal });
+            while (!poller.isDone) {
+              await delay(currentPollIntervalInMs, { abortSignal });
+              await poller.poll({ abortSignal });
+            }
+          }
+        } finally {
+          inputAbortSignal?.removeEventListener("abort", abortListener);
+        }
+        switch (state.status) {
+          case "succeeded":
+            return poller.result;
+          case "canceled":
+            throw new Error(cancelErrMsg);
+          case "failed":
+            throw state.error;
+          case "notStarted":
+          case "running":
+            throw new Error(`Polling completed without succeeding or failing`);
+        }
+      })().finally(() => {
```

Comment:
nit: prefer avoiding mixing async/await with promise chaining if at all possible? It can get confusing but I understand if that's not realistic here

Reply:
I don't have better ideas here, but I am open to any. Chaining is used because the promise needs to be stored and async/await is used to simplify the meat of the first promise.